File: app/backend/controllers/iop/transfer/impl/ExternalCalendarController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package backend.controllers.iop.transfer.impl;

import backend.controllers.CalendarController;
import backend.exceptions.NotFoundException;
import backend.impl.CalendarHandler;
import backend.models.Exam;
import backend.models.ExamEnrolment;
import backend.models.ExamMachine;
import backend.models.ExamRoom;
import backend.models.Reservation;
import backend.models.User;
import backend.sanitizers.Attrs;
import backend.sanitizers.ExternalCalendarReservationSanitizer;
import backend.security.Authenticated;
import backend.util.config.ConfigReader;
import backend.util.datetime.DateTimeUtils;
import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import be.objectify.deadbolt.java.actions.SubjectNotPresent;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.typesafe.config.ConfigFactory;
import io.ebean.Ebean;
import io.ebean.text.PathProperties;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import java.util.stream.Collector;
import javax.inject.Inject;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.Interval;
import org.joda.time.LocalDate;
import org.joda.time.format.ISODateTimeFormat;
import play.libs.Json;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;

public class ExternalCalendarController extends CalendarController {
    @Inject
    private WSClient wsClient;

    @Inject
    private CalendarHandler calendarHandler;

    @Inject
    private ConfigReader configReader;

    private static URL parseUrl(String orgRef, String facilityRef, String date, String start, String end, int duration)
        throws MalformedURLException {
        String url =
            ConfigFactory.load().getString("sitnet.integration.iop.host") +
            String.format("/api/organisations/%s/facilities/%s/slots", orgRef, facilityRef) +
            String.format("?date=%s&startAt=%s&endAt=%s&duration=%d", date, start, end, duration);
        return new URL(url);
    }

    private static URL parseUrl(String orgRef, String facilityRef) throws MalformedURLException {
        return new URL(
            ConfigFactory.load().getString("sitnet.integration.iop.host") +
            String.format("/api/organisations/%s/facilities/%s/reservations", orgRef, facilityRef)
        );
    }

    private static URL parseUrl(String orgRef, String facilityRef, String reservationRef) throws MalformedURLException {
        return new URL(
            ConfigFactory.load().getString("sitnet.integration.iop.host") +
            String.format(
                "/api/organisations/%s/facilities/%s/reservations/%s/force",
                orgRef,
                facilityRef,
                reservationRef
            )
        );
    }

    // Actions invoked by central IOP server

    @SubjectNotPresent
    public Result provideReservation(Http.Request request) {
        // Parse request body
        JsonNode node = request.body().asJson();
        String reservationRef = node.get("id").asText();
        String roomRef = node.get("roomId").asText();
        DateTime start = ISODateTimeFormat.dateTimeParser().parseDateTime(node.get("start").asText());
        DateTime end = ISODateTimeFormat.dateTimeParser().parseDateTime(node.get("end").asText());
        String userEppn = node.get("user").asText();
        if (start.isBeforeNow() || end.isBefore(start)) {
            return badRequest("invalid dates");
        }
        ExamRoom room = Ebean.find(ExamRoom.class).where().eq("externalRef", roomRef).findOne();
        if (room == null) {
            return notFound("room not found");
        }
        Optional<ExamMachine> machine = calendarHandler.getRandomMachine(
            room,
            null,
            start,
            end,
            Collections.emptyList()
        );
        if (machine.isEmpty()) {
            return forbidden("sitnet_no_machines_available");
        }
        // We are good to go :)
        Reservation reservation = new Reservation();
        reservation.setExternalRef(reservationRef);
        reservation.setEndAt(end);
        reservation.setStartAt(start);
        reservation.setMachine(machine.get());
        reservation.setExternalUserRef(userEppn);
        reservation.save();
        PathProperties pp = PathProperties.parse("(*, machine(*, room(*, mailAddress(*))))");

        return created(reservation, pp);
    }

    // Initiated by originator of reservation (the student)
    @SubjectNotPresent
    public Result acknowledgeReservationRemoval(String ref) {
        Reservation reservation = Ebean
            .find(Reservation.class)
            .fetch("machine")
            .fetch("machine.room")
            .where()
            .eq("externalRef", ref)
            .findOne();
        if (reservation == null) {
            return notFound("reservation not found");
        }
        // TODO: might need additional checks
        DateTime now = DateTimeUtils.adjustDST(DateTime.now(), reservation);
        if (reservation.toInterval().isBefore(now) || reservation.toInterval().contains(now)) {
            return forbidden("sitnet_reservation_in_effect");
        }
        if (reservation.getEnrolment() != null) {
            reservation.getEnrolment().delete(); // cascades to reservation
        } else {
            reservation.delete();
        }
        return ok();
    }

    // Initiated by administrator of organisation where reservation takes place
    @SubjectNotPresent
    public Result acknowledgeReservationRevocation(String ref) {
        ExamEnrolment enrolment = Ebean
            .find(ExamEnrolment.class)
            .fetch("reservation")
            .fetch("reservation.externalReservation")
            .fetch("reservation.machine")
            .fetch("reservation.machine.room")
            .where()
            .eq("reservation.externalRef", ref)
            .findOne();
        if (enrolment == null) {
            return notFound(String.format("No reservation with ref %s.", ref));
        }

        Reservation reservation = enrolment.getReservation();
        DateTime now = DateTimeUtils.adjustDST(DateTime.now(), reservation);
        if (reservation.toInterval().isBefore(now) || reservation.toInterval().contains(now)) {
            return forbidden("sitnet_reservation_in_effect");
        }
        enrolment.setReservation(null);
        enrolment.update();
        reservation.delete();
        return ok();
    }

    @SubjectNotPresent
    public Result provideSlots(
        Optional<String> roomId,
        Optional<String> date,
        Optional<String> start,
        Optional<String> end,
        Optional<Integer> duration
    ) {
        if (roomId.isPresent() && date.isPresent() && start.isPresent() && end.isPresent() && duration.isPresent()) {
            ExamRoom room = Ebean.find(ExamRoom.class).where().eq("externalRef", roomId.get()).findOne();
            if (room == null) {
                return forbidden(String.format("No room with ref: (%s)", roomId.get()));
            }
            Collection<CalendarHandler.TimeSlot> slots = new ArrayList<>();
            if (!room.getOutOfService() && !room.getState().equals(ExamRoom.State.INACTIVE.toString())) {
                LocalDate searchDate;
                try {
                    searchDate = parseSearchDate(date.get(), start.get(), end.get(), room);
                } catch (NotFoundException e) {
                    return notFound();
                }
                List<ExamMachine> machines = Ebean
                    .find(ExamMachine.class)
                    .where()
                    .eq("room.id", room.getId())
                    .ne("outOfService", true)
                    .ne("archived", true)
                    .findList();
                LocalDate endOfSearch = getEndSearchDate(end.get(), searchDate);
                while (!searchDate.isAfter(endOfSearch)) {
                    Set<CalendarHandler.TimeSlot> timeSlots = getExamSlots(room, duration.get(), searchDate, machines);
                    if (!timeSlots.isEmpty()) {
                        slots.addAll(timeSlots);
                    }
                    searchDate = searchDate.plusDays(1);
                }
            }
            return ok(Json.toJson(slots));
        } else {
            return badRequest();
        }
    }

    // Actions invoked directly by logged in users
    @Authenticated
    @With(ExternalCalendarReservationSanitizer.class)
    @Restrict(@Group("STUDENT"))
    public CompletionStage<Result> requestReservation(Http.Request request) throws MalformedURLException {
        if (!configReader.isVisitingExaminationSupported()) {
            return wrapAsPromise(forbidden("Feature not enabled in the installation"));
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        // Parse request body
        String orgRef = request.attrs().get(Attrs.ORG_REF);
        String roomRef = request.attrs().get(Attrs.ROOM_REF);
        DateTime start = request.attrs().get(Attrs.START_DATE);
        DateTime end = request.attrs().get(Attrs.END_DATE);
        Long examId = request.attrs().get(Attrs.EXAM_ID);
        Collection<Long> sectionIds = request.attrs().get(Attrs.SECTION_IDS);

        //TODO: See if this offset thing works as intended
        DateTime now = DateTimeUtils.adjustDST(DateTime.now());
        Optional<ExamEnrolment> oe = Ebean
            .find(ExamEnrolment.class)
            .fetch("reservation")
            .fetch("exam.examSections")
            .fetch("exam.examSections.examMaterials")
            .where()
            .eq("user.id", user.getId())
            .eq("exam.id", examId)
            .eq("exam.state", Exam.State.PUBLISHED)
            .or()
            .isNull("reservation")
            .gt("reservation.startAt", now.toDate())
            .endOr()
            .findOneOrEmpty();
        if (oe.isEmpty()) {
            return wrapAsPromise(forbidden("sitnet_error_enrolment_not_found"));
        }
        ExamEnrolment enrolment = oe.get();
        Optional<Result> error = checkEnrolment(enrolment, user, sectionIds);
        if (error.isPresent()) {
            return wrapAsPromise(error.get());
        }
        // Lets do this
        URL url = parseUrl(orgRef, roomRef);
        String homeOrgRef = ConfigFactory.load().getString("sitnet.integration.iop.organisationRef");
        ObjectNode body = Json.newObject();
        body.put("requestingOrg", homeOrgRef);
        body.put("start", ISODateTimeFormat.dateTime().print(start));
        body.put("end", ISODateTimeFormat.dateTime().print(end));
        body.put("user", user.getEppn());
        body.set(
            "optionalSections",
            sectionIds.stream().collect(Collector.of(Json::newArray, ArrayNode::add, ArrayNode::add))
        );

        WSRequest wsRequest = wsClient.url(url.toString());
        return wsRequest
            .post(body)
            .thenComposeAsync(
                response -> {
                    JsonNode root = response.asJson();
                    if (response.getStatus() != Http.Status.CREATED) {
                        return wrapAsPromise(internalServerError(root.get("message").asText("Connection refused")));
                    }
                    return calendarHandler
                        .handleExternalReservation(
                            enrolment,
                            enrolment.getExam(),
                            root,
                            start,
                            end,
                            user,
                            orgRef,
                            roomRef,
                            sectionIds
                        )
                        .thenApplyAsync(
                            err -> {
                                if (err.isEmpty()) {
                                    return created(root.get("id"));
                                }
                                return internalServerError();
                            }
                        );
                }
            );
    }

    @Authenticated
    @Restrict(@Group("STUDENT"))
    public CompletionStage<Result> requestReservationRemoval(String ref, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        Reservation reservation = Ebean.find(Reservation.class).where().eq("externalRef", ref).findOne();
        return externalReservationHandler.removeReservation(reservation, user, "");
    }

    @Restrict(@Group("ADMIN"))
    public CompletionStage<Result> requestReservationRevocation(String ref, Http.Request request)
        throws MalformedURLException {
        Optional<Reservation> or = Ebean
            .find(Reservation.class)
            .where()
            .isNotNull("machine")
            .eq("externalRef", ref)
            .isNull("enrolment")
            .findOneOrEmpty();
        if (or.isEmpty()) {
            return CompletableFuture.completedFuture(notFound(String.format("No reservation with ref %s.", ref)));
        }

        Reservation reservation = or.get();
        DateTime now = DateTimeUtils.adjustDST(DateTime.now(), reservation);
        if (reservation.toInterval().isBefore(now) || reservation.toInterval().contains(now)) {
            return CompletableFuture.completedFuture(forbidden("sitnet_reservation_in_effect"));
        }
        String roomRef = reservation.getMachine().getRoom().getExternalRef();
        URL url = parseUrl(configReader.getHomeOrganisationRef(), roomRef, reservation.getExternalRef());
        WSRequest wsRequest = wsClient.url(url.toString());

        Function<WSResponse, Result> onSuccess = response -> {
            if (response.getStatus() != Http.Status.OK) {
                JsonNode root = response.asJson();
                return internalServerError(root.get("message").asText("Connection refused"));
            }
            String msg = request.body().asJson().path("msg").asText("");
            emailComposer.composeExternalReservationCancellationNotification(reservation, msg);
            reservation.delete();
            return ok();
        };

        return wsRequest.delete().thenApplyAsync(onSuccess);
    }

    @Authenticated
    @Restrict(@Group("STUDENT"))
    public CompletionStage<Result> requestSlots(
        Long examId,
        String roomRef,
        Optional<String> org,
        Optional<String> date,
        Http.Request request
    )
        throws MalformedURLException {
        if (org.isPresent() && date.isPresent()) {
            // First check that exam exists
            User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
            ExamEnrolment ee = getEnrolment(examId, user);
            // For now do not allow making an external reservation for collaborative exam
            if (ee == null || ee.getCollaborativeExam() != null) {
                return wrapAsPromise(forbidden("sitnet_error_enrolment_not_found"));
            }
            Exam exam = ee.getExam();

            // Also sanity check the provided search date
            try {
                calendarHandler.parseSearchDate(date.get(), exam, null);
            } catch (NotFoundException e) {
                return wrapAsPromise(notFound());
            }
            // Ready to shoot
            String start = ISODateTimeFormat.dateTime().print(new DateTime(exam.getExamActiveStartDate()));
            String end = ISODateTimeFormat.dateTime().print(new DateTime(exam.getExamActiveEndDate()));
            Integer duration = exam.getDuration();
            URL url = parseUrl(org.get(), roomRef, date.get(), start, end, duration);
            WSRequest wsRequest = wsClient.url(url.toString().split("\\?")[0]).setQueryString(url.getQuery());
            Function<WSResponse, Result> onSuccess = response -> {
                JsonNode root = response.asJson();
                if (response.getStatus() != Http.Status.OK) {
                    return internalServerError(root.get("message").asText("Connection refused"));
                }
                Set<CalendarHandler.TimeSlot> slots = calendarHandler.postProcessSlots(root, date.get(), exam, user);
                return ok(Json.toJson(slots));
            };
            return wsRequest.get().thenApplyAsync(onSuccess);
        } else {
            return wrapAsPromise(badRequest());
        }
    }

    // helpers ->

    private Set<CalendarHandler.TimeSlot> getExamSlots(
        ExamRoom room,
        Integer examDuration,
        LocalDate date,
        Collection<ExamMachine> machines
    ) {
        Set<CalendarHandler.TimeSlot> slots = new LinkedHashSet<>();
        Collection<Interval> examSlots = calendarHandler.gatherSuitableSlots(room, date, examDuration);
        // Check machine availability for each slot
        for (Interval slot : examSlots) {
            // Check machine availability
            int availableMachineCount = (int) machines.stream().filter(m -> !isReservedDuring(m, slot)).count();
            slots.add(new CalendarHandler.TimeSlot(slot, availableMachineCount, null));
        }
        return slots;
    }

    /**
     * Search date is the current date if searching for current week or earlier,
     * If searching for upcoming weeks, day of week is one.
     */
    private LocalDate parseSearchDate(String day, String startDate, String endDate, ExamRoom room)
        throws NotFoundException {
        int windowSize = calendarHandler.getReservationWindowSize();
        int offset = room != null
            ? DateTimeZone.forID(room.getLocalTimezone()).getOffset(DateTime.now())
            : configReader.getDefaultTimeZone().getOffset(DateTime.now());
        LocalDate now = DateTime.now().plusMillis(offset).toLocalDate();
        LocalDate reservationWindowDate = now.plusDays(windowSize);
        LocalDate examEndDate = DateTime
            .parse(endDate, ISODateTimeFormat.dateTimeParser())
            .plusMillis(offset)
            .toLocalDate();
        LocalDate searchEndDate = reservationWindowDate.isBefore(examEndDate) ? reservationWindowDate : examEndDate;

        LocalDate examStartDate = DateTime
            .parse(startDate, ISODateTimeFormat.dateTimeParser())
            .plusMillis(offset)
            .toLocalDate();
        LocalDate searchDate = day.equals("") ? now : LocalDate.parse(day, ISODateTimeFormat.dateParser());
        searchDate = searchDate.withDayOfWeek(1);
        if (searchDate.isBefore(now)) {
            searchDate = now;
        }
        // if searching for month(s) after exam's end month -> no can do
        if (searchDate.isAfter(searchEndDate)) {
            throw new NotFoundException();
        }
        // Do not execute search before exam starts
        if (searchDate.isBefore(examStartDate)) {
            searchDate = examStartDate;
        }
        return searchDate;
    }

    /**
     * @return which one is sooner, exam period's end or week's end
     */
    private LocalDate getEndSearchDate(String endDate, LocalDate searchDate) {
        LocalDate examEnd = LocalDate.parse(endDate, ISODateTimeFormat.dateTimeParser());
        return calendarHandler.getEndSearchDate(searchDate, examEnd);
    }

    private boolean isReservedDuring(ExamMachine machine, Interval interval) {
        return machine.getReservations().stream().anyMatch(r -> interval.overlaps(r.toInterval()));
    }
}

File: app/features/iop/transfer/controllers/ExternalCalendarController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package features.iop.transfer.controllers

import database.{EbeanJsonExtensions, EbeanQueryExtensions}
import features.iop.transfer.services.ExternalReservationHandlerService
import io.ebean.DB
import models.calendar.MaintenancePeriod
import models.enrolment.{ExamEnrolment, Reservation}
import models.exam.ExamState
import models.facility.{ExamMachine, ExamRoom}
import models.user.Role
import play.api.Logging
import play.api.libs.json.{JsValue, Json, Writes}
import play.api.libs.ws.{JsonBodyWritables, WSClient}
import play.api.mvc.*
import security.Auth.{AuthenticatedAction, authorized, subjectNotPresent}
import security.{Auth, BlockingIOExecutionContext}
import services.config.ConfigReader
import services.datetime.Interval
import services.datetime.IntervalExtensions.*
import services.datetime.{AppClock, CalendarHandler, TimeUtils}
import services.mail.EmailComposer
import system.AuditedAction
import validation.calendar.ExternalCalendarReservationValidator

import java.net.{MalformedURLException, URI, URL}
import java.time.*
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters
import javax.inject.Inject
import scala.concurrent.Future
import scala.jdk.CollectionConverters.*
import scala.util.Try

class ExternalCalendarController @Inject() (
    authenticated: AuthenticatedAction,
    audited: AuditedAction,
    wsClient: WSClient,
    calendarHandler: CalendarHandler,
    emailComposer: EmailComposer,
    configReader: ConfigReader,
    externalReservationHandler: ExternalReservationHandlerService,
    clock: AppClock,
    val controllerComponents: ControllerComponents,
    implicit val ec: BlockingIOExecutionContext
) extends BaseController
    with EbeanQueryExtensions
    with EbeanJsonExtensions
    with JsonBodyWritables
    with Logging:

  private def parseUrl(
      orgRef: String,
      facilityRef: String,
      date: String,
      start: String,
      end: String,
      duration: Int
  ): URL =
    val url = configReader.getIopHost +
      s"/api/organisations/$orgRef/facilities/$facilityRef/slots" +
      s"?date=$date&startAt=$start&endAt=$end&duration=$duration"
    URI.create(url).toURL

  private def parseUrl(orgRef: String, facilityRef: String): URL =
    URI
      .create(
        configReader.getIopHost +
          s"/api/organisations/$orgRef/facilities/$facilityRef/reservations"
      )
      .toURL

  private def parseUrl(orgRef: String, facilityRef: String, reservationRef: String): URL =
    URI
      .create(
        configReader.getIopHost +
          s"/api/organisations/$orgRef/facilities/$facilityRef/reservations/$reservationRef/force"
      )
      .toURL

  // Actions invoked by central IOP server

  def provideReservation(): Action[JsValue] = audited.andThen(subjectNotPresent)(parse.json) {
    request =>
      // Parse request body
      val body           = request.body
      val reservationRef = (body \ "id").as[String]
      val roomRef        = (body \ "roomId").as[String]
      val start          = TimeUtils.parseInstant((body \ "start").as[String])
      val end            = TimeUtils.parseInstant((body \ "end").as[String])
      val userEppn       = (body \ "user").as[String]
      val orgRef         = (body \ "orgRef").as[String]
      val orgName        = (body \ "orgName").as[String]

      if start.isBefore(clock.now()) || end.isBefore(start) then
        BadRequest("invalid dates")
      else
        DB.find(classOf[ExamRoom]).where().eq("externalRef", roomRef).find match
          case None => NotFound("room not found")
          case Some(room) =>
            calendarHandler.getRandomMachine(room, null, start, end, Seq.empty) match
              case None          => Forbidden("i18n_no_machines_available")
              case Some(machine) =>
                // We are good to go :)
                val reservation = new Reservation()
                reservation.externalRef = reservationRef
                reservation.endAt = end
                reservation.startAt = start
                reservation.machine = machine
                reservation.externalUserRef = userEppn
                reservation.externalOrgRef = orgRef
                reservation.externalOrgName = orgName
                reservation.save()

                Created(asJson(reservation))
  }

  // Initiated by originator of reservation (the student)
  def acknowledgeReservationRemoval(ref: String): Action[AnyContent] = Action { _ =>
    DB.find(classOf[Reservation])
      .fetch("machine")
      .fetch("machine.room")
      .fetch("enrolment")
      .where()
      .eq("externalRef", ref)
      .find match
      case None => NotFound("reservation not found")
      case Some(reservation) =>
        val now = clock.now()
        if reservation.toInterval.isBefore(now) || reservation.toInterval.contains(now) then
          Forbidden("i18n_reservation_in_effect")
        else
          Option(reservation.enrolment) match
            case Some(enrolment) => enrolment.delete() // cascades to reservation
            case None            => reservation.delete()
          Ok
  }

  // Initiated by administrator of organization where reservation takes place
  def acknowledgeReservationRevocation(ref: String): Action[AnyContent] = Action { _ =>
    DB.find(classOf[ExamEnrolment])
      .fetch("reservation")
      .fetch("reservation.externalReservation")
      .fetch("reservation.machine")
      .fetch("reservation.machine.room")
      .where()
      .eq("reservation.externalRef", ref)
      .find match
      case None => NotFound(f"No reservation with ref $ref.")
      case Some(enrolment) =>
        val reservation = enrolment.reservation
        val now         = clock.now()
        if reservation.toInterval.isBefore(now) || reservation.toInterval.contains(now) then
          Forbidden("i18n_reservation_in_effect")
        else
          enrolment.reservation = null
          enrolment.update()
          reservation.delete()
          Ok
  }

  def provideSlots(
      roomId: Option[String],
      date: Option[String],
      start: Option[String],
      end: Option[String],
      duration: Option[Int]
  ): Action[AnyContent] = Action { _ =>
    (roomId, date, start, end, duration) match
      case (Some(rid), Some(d), Some(s), Some(e), Some(dur)) =>
        DB.find(classOf[ExamRoom]).where().eq("externalRef", rid).find match
          case None => Forbidden(f"No room with ref: ($rid)")
          case Some(room) =>
            val slots =
              if !room.outOfService && room.state != ExamRoom.State.INACTIVE.toString then
                parseSearchDate(d, s, e, room) match
                  case None => List.empty[CalendarHandler.TimeSlot]
                  case Some(searchDate) =>
                    val machines = DB
                      .find(classOf[ExamMachine])
                      .where()
                      .eq("room.id", room.id)
                      .ne("outOfService", true)
                      .ne("archived", true)
                      .list

                    // Maintenance periods
                    val periods = DB
                      .find(classOf[MaintenancePeriod])
                      .where()
                      .gt("endsAt", searchDate.atStartOfDay(ZoneOffset.UTC).toInstant)
                      .list
                      .map(p => p.startsAt to p.endsAt)

                    val endOfSearch = getEndSearchDate(e, searchDate)

                    Iterator
                      .iterate(searchDate)(_.plusDays(1))
                      .takeWhile(!_.isAfter(endOfSearch))
                      .flatMap { date =>
                        getExamSlots(room, dur, date, machines, periods)
                      }
                      .toList
              else List.empty

            implicit val timeSlotWrites: Writes[CalendarHandler.TimeSlot] =
              (slot: CalendarHandler.TimeSlot) =>
                Json.obj(
                  "start"             -> slot.start,
                  "end"               -> slot.end,
                  "availableMachines" -> slot.availableMachines,
                  "ownReservation"    -> slot.ownReservation,
                  "conflictingExam"   -> slot.conflictingExam
                )

            Ok(Json.toJson(slots))
      case _ => BadRequest("Missing required parameters")
  }

  // Actions invoked directly by logged-in users
  def requestReservation(): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.STUDENT)))
    .andThen(ExternalCalendarReservationValidator.filter(clock))
    .async(parse.json) { request =>
      if !configReader.isVisitingExaminationSupported then
        Future.successful(Forbidden("Feature not enabled in the installation"))
      else
        val user       = request.attrs(Auth.ATTR_USER)
        val dto        = request.attrs(validation.core.ScalaAttrs.ATTR_EXT_RESERVATION)
        val orgRef     = dto.orgRef
        val roomRef    = dto.roomRef
        val start      = dto.start
        val end        = dto.end
        val examId     = dto.examId
        val sectionIds = dto.sectionIds

        val enrolmentOpt =
          DB.find(classOf[ExamEnrolment])
            .fetch("reservation")
            .fetch("exam.examSections")
            .fetch("exam.examSections.examMaterials")
            .where()
            .eq("user.id", user.id)
            .eq("exam.id", examId)
            .eq("exam.state", ExamState.PUBLISHED)
            .disjunction()
            .isNull("reservation")
            .gt("reservation.startAt", clock.now())
            .endJunction()
            .find

        enrolmentOpt match
          case None => Future.successful(Forbidden("i18n_error_enrolment_not_found"))
          case Some(enrolment) =>
            val sectionIdsSeq = sectionIds.getOrElse(List.empty)
            calendarHandler.checkEnrolment(enrolment, user, sectionIdsSeq) match
              case Some(errorResult) => Future.successful(errorResult)
              case None              =>
                // Let's do this
                Try(parseUrl(orgRef, roomRef)).fold(
                  {
                    case e: MalformedURLException =>
                      logger.error("Invalid URL", e)
                      Future.successful(BadRequest("Invalid URL"))
                    case e => throw e
                  },
                  url =>
                    val homeOrgRef = configReader.getHomeOrganisationRef
                    val body = Json.obj(
                      "requestingOrg"    -> homeOrgRef,
                      "start"            -> DateTimeFormatter.ISO_INSTANT.format(start),
                      "end"              -> DateTimeFormatter.ISO_INSTANT.format(end),
                      "user"             -> user.eppn,
                      "optionalSections" -> Json.toJson(sectionIdsSeq)
                    )

                    for
                      response <- wsClient.url(url.toString).post(body)
                      result <-
                        if response.status != CREATED then
                          val root = response.json.as[play.api.libs.json.JsObject]
                          val msg = (root \ "message").asOpt[String].getOrElse("Connection refused")
                          Future.successful(InternalServerError(msg))
                        else
                          val root = response.json
                          calendarHandler
                            .handleExternalReservation(
                              enrolment,
                              enrolment.exam,
                              root,
                              start,
                              end,
                              user,
                              orgRef,
                              roomRef,
                              sectionIdsSeq
                            )
                            .map {
                              case Some(_) => InternalServerError("Internal server error")
                              case None    => Created((root \ "id").get)
                            }
                    yield result
                )
    }

  def requestReservationRemoval(ref: String): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.STUDENT))).async { request =>
      val user = request.attrs(Auth.ATTR_USER)
      DB.find(classOf[Reservation]).where().eq("externalRef", ref).find match
        case None => Future.successful(NotFound(f"No reservation with ref $ref."))
        case Some(reservation) =>
          externalReservationHandler.removeReservation(reservation, user, "").map(_ => Ok)
    }

  def requestReservationRevocation(ref: String): Action[AnyContent] =
    Action.andThen(authorized(Seq(Role.Name.ADMIN))).async { request =>
      val reservationOpt =
        DB.find(classOf[Reservation])
          .where()
          .isNotNull("machine")
          .eq("externalRef", ref)
          .isNull("enrolment")
          .find

      reservationOpt match
        case None => Future.successful(NotFound(f"No reservation with ref $ref."))
        case Some(reservation) =>
          val now = clock.now()
          if reservation.toInterval.isBefore(now) || reservation.toInterval.contains(now) then
            Future.successful(Forbidden("i18n_reservation_in_effect"))
          else
            val roomRef = reservation.machine.room.externalRef
            Try(parseUrl(configReader.getHomeOrganisationRef, roomRef, reservation.externalRef))
              .fold(
                {
                  case e: MalformedURLException =>
                    logger.error("Invalid URL", e)
                    Future.successful(BadRequest("Invalid URL"))
                  case e => throw e
                },
                url =>
                  wsClient
                    .url(url.toString)
                    .delete()
                    .map { response =>
                      if response.status != OK then
                        val root = response.json.as[play.api.libs.json.JsObject]
                        val msg  = (root \ "message").asOpt[String].getOrElse("Connection refused")
                        InternalServerError(msg)
                      else
                        val msg =
                          request.body.asJson.map(_ \ "msg").flatMap(_.asOpt[String]).getOrElse("")
                        emailComposer.composeExternalReservationCancellationNotification(
                          reservation,
                          Some(msg)
                        )
                        reservation.delete()
                        Ok
                    }
              )
    }

  def requestSlots(
      examId: Long,
      roomRef: String,
      org: Option[String],
      date: Option[String]
  ): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.STUDENT))).async { request =>
      (org, date) match
        case (Some(orgRef), Some(d)) =>
          val user = request.attrs(Auth.ATTR_USER)
          Option(calendarHandler.getEnrolment(examId, user)) match
            case None => Future.successful(Forbidden("i18n_error_enrolment_not_found"))
            case Some(ee) if Option(ee.collaborativeExam).isDefined =>
              Future.successful(Forbidden("i18n_error_enrolment_not_found"))
            case Some(ee) =>
              val exam = ee.exam

              // Sanity-check the provided search date
              calendarHandler.parseSearchDate(d, exam, None) match
                case None => Future.successful(NotFound("Invalid search date"))
                case _    =>
                  // Ready to shoot
                  val start    = DateTimeFormatter.ISO_INSTANT.format(exam.periodStart)
                  val end      = DateTimeFormatter.ISO_INSTANT.format(exam.periodEnd)
                  val duration = exam.duration
                  Try(parseUrl(orgRef, roomRef, d, start, end, duration)).fold(
                    {
                      case e: MalformedURLException =>
                        logger.error("Invalid URL", e)
                        Future.successful(BadRequest("Invalid URL"))
                      case e => throw e
                    },
                    url =>
                      wsClient
                        .url(url.toString.split("\\?")(0))
                        .withQueryStringParameters(url.getQuery.split("&").map { kv =>
                          val parts = kv.split("=")
                          parts(0) -> parts(1)
                        }*)
                        .get()
                        .map { response =>
                          if response.status != OK then
                            val root = response.json.as[play.api.libs.json.JsObject]
                            val msg =
                              (root \ "message").asOpt[String].getOrElse("Connection refused")
                            InternalServerError(msg)
                          else
                            val root  = response.json
                            val slots = calendarHandler.postProcessSlots(root, d, exam, user)
                            Ok(Json.toJson(slots))
                        }
                  )
        case _ => Future.successful(BadRequest("Missing required parameters"))
    }

  // Helpers

  private def getExamSlots(
      room: ExamRoom,
      examDuration: Int,
      date: LocalDate,
      machines: List[ExamMachine],
      maintenances: List[Interval]
  ): Set[CalendarHandler.TimeSlot] =
    val examSlots = calendarHandler
      .gatherSuitableSlots(room, date, examDuration)
      .filterNot(slot => maintenances.exists(_.overlaps(slot)))

    // Check machine availability for each slot
    examSlots.map { slot =>
      val availableMachineCount = machines.count(m => !isReservedDuring(m, slot))
      CalendarHandler.TimeSlot(slot, availableMachineCount, null)
    }.toSet

  private def parseSearchDate(
      day: String,
      startDate: String,
      endDate: String,
      room: ExamRoom
  ): Option[LocalDate] =
    val windowSize = calendarHandler.getReservationWindowSize
    val zoneId = Option(room)
      .map(r => TimeUtils.zoneIdOf(r.localTimezone))
      .getOrElse(ZoneId.of(configReader.getDefaultTimeZone.getId))

    val now                   = clock.now().atZone(zoneId).toLocalDate
    val reservationWindowDate = now.plusDays(windowSize)
    val examEndDate           = TimeUtils.parseInstant(endDate).atZone(zoneId).toLocalDate
    val searchEndDate =
      if reservationWindowDate.isBefore(examEndDate) then reservationWindowDate else examEndDate

    val examStartDate = TimeUtils.parseInstant(startDate).atZone(zoneId).toLocalDate

    val initialDate =
      if day.isEmpty then now else LocalDate.parse(day)
    val weekStart = initialDate.`with`(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
    val afterNow  = if weekStart.isBefore(now) then now else weekStart

    if afterNow.isAfter(searchEndDate) then None
    else
      // Do not execute search before exam starts
      val searchDate = if afterNow.isBefore(examStartDate) then examStartDate else afterNow
      Some(searchDate)

  private def getEndSearchDate(endDate: String, searchDate: LocalDate): LocalDate =
    val examEnd = TimeUtils.parseInstant(endDate).atZone(ZoneOffset.UTC).toLocalDate
    calendarHandler.getEndSearchDate(searchDate, examEnd)

  private def isReservedDuring(machine: ExamMachine, interval: Interval): Boolean =
    machine.reservations.asScala.exists(r => interval.overlaps(r.toInterval))
